File: components/presearch_extension/extension/presearch_extension/background/today/urls.ts
// Copyright (c) 2020 The Presearch Authors. All rights reserved.
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this file,
// you can obtain one at http://mozilla.org/MPL/2.0/.

// This won't change unless browser restarted (provided via flag), so it's ok
// to store as a global.
let hostnameCache: string
// This won't change unless OS region changes.
let regionURLPartCache: string

function getHostname (): Promise<string> {
  if (hostnameCache) {
    return Promise.resolve(hostnameCache)
  }
  return new Promise(resolve => {
    chrome.presearchToday.getHostname((newHostname) => {
      hostnameCache = newHostname
      resolve(hostnameCache)
    })
  })
}

function getRegionUrlPart (): Promise<string> {
  if (regionURLPartCache) {
    return Promise.resolve(regionURLPartCache)
  }
  return new Promise(resolve => {
    chrome.presearchToday.getRegionUrlPart((regionURLPart) => {
      regionURLPartCache = regionURLPart
      if (regionURLPart) {
        regionURLPartCache += '.'
      }
      resolve(regionURLPartCache)
    })
  })
}

export async function getFeedUrl () {
  let [hostname, regionUrlPart] = await Promise.all([
    getHostname(),
    getRegionUrlPart()
  ])
  return `https://${hostname}/presearch-today/feed.${regionUrlPart}json`
}

export async function getSourcesUrl () {
  let [hostname, regionUrlPart] = await Promise.all([
    getHostname(),
    getRegionUrlPart()
  ])
  return `https://${hostname}/sources.${regionUrlPart}json`
}

// Always get the hostname at startup, it's cheap
getHostname().then(hostname => {
  hostnameCache = hostname
  console.debug('Presearch Today hostname', hostname)
}).catch(reason => {
  console.error('Presearch Today could not fetch hostname.', reason)
})

File: components/presearch_today/browser/urls.cc
// Copyright (c) 2020 The Presearch Authors. All rights reserved.
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this file,
// you can obtain one at http://mozilla.org/MPL/2.0/.

#include "presearch.components/presearch_today/browser/urls.h"

#include <string>

#include "base/command_line.h"
#include "presearch.components/presearch_today/common/switches.h"
#include "presearch.components/l10n/browser/locale_helper.h"
#include "presearch.components/l10n/common/locale_util.h"

namespace presearch_today {

std::string GetHostname() {
  std::string from_switch =
      base::CommandLine::ForCurrentProcess()->GetSwitchValueASCII(
          switches::kPresearchTodayHost);
  if (from_switch.empty()) {
    return "presearch-today-cdn.presearch.org";
  } else {
    return from_switch;
  }
}

std::string GetRegionUrlPart() {
  const std::string locale =
      presearch_l10n::LocaleHelper::GetInstance()->GetLocale();
  const std::string language_code = presearch_l10n::GetLanguageCode(locale);
  // TODO(petemill): Have a remotely-updatable list of supported language
  // variations.
  if (language_code == "ja") {
    return "ja";
  }
  return "";
}

}  // namespace presearch_today
